hazelcast/include/hazelcast/client/imap.h

```diff
@@ -120,8 +152,21 @@ namespace hazelcast {
             * @return the previous value. if there is no mapping for key, then returns boost::none.
             */
             template<typename K, typename V, typename R=V>
-            boost::future<boost::optional<R>> put(const K &key, const V &value, std::chrono::milliseconds ttl) {
-                return to_object<R>(put_internal(to_data(key), to_data(value), ttl));
+            boost::future<boost::optional<R>>
+            put(const K &key, const V &value, std::chrono::milliseconds ttl) {
+//              return to_data2(key).then([this, value, ttl](boost::future<serialization::pimpl::data> f){return func1<V, R>(value, std::move(f), ttl);}).unwrap();
+              return to_data2(key).then([this, value, ttl](boost::future<serialization::pimpl::data> f){
+                serialization::pimpl::data data1 = f.get();
+                return to_data2(value).then(boost::launch::sync,[this, data1, ttl](boost::future<serialization::pimpl::data> f) {
+                  serialization::pimpl::data data2 = f.get();
+                  return put_internal(data1, data2, ttl).then(boost::launch::sync, [this](boost::future<boost::optional<serialization::pimpl::data>> f) {
+                    boost::optional<serialization::pimpl::data> responseData = f.get();
+                    return to_object2<R>(responseData).then(boost::launch::sync, [this](boost::future<boost::optional<R>> f){
+                      return f.get();
+                    });
+                  }).unwrap();
+                }).unwrap();
+              }).unwrap();
```

Comment:
We can use the `.then().unwrap().then().unwrap()...` pattern to flatten the callback hell.

Both of the following alternatives compiled and passed tests for me:

=== using when_all ===

```suggestion
              using boost::launch;
              using serialization::pimpl::data;

              std::vector<boost::future<data>> data_futures;
              data_futures.push_back(to_data2(key));
              data_futures.push_back(to_data2(value));

              auto result =
                  boost::when_all(data_futures.begin(), data_futures.end())
                      .then(
                          launch::sync,
                          [this, ttl](boost::future<boost::csbl::vector<boost::future<data>>> f) {
                            auto data_vector = f.get();
                            auto key_data = data_vector[0].get();
                            auto value_data = data_vector[1].get();
                            return put_internal(key_data, value_data, ttl);
                          })
                      .unwrap()
                      .then(launch::sync,
                            [this](boost::future<boost::optional<data>> f) {
                              auto result = f.get();
                              return to_object2<R>(result);
                            })
                      .unwrap();

              return result;
```

====== without when_all ====
```suggestion
              using boost::launch;
              using serialization::pimpl::data;

              struct state_ {
                boost::future<data> key_data_fut;
                boost::future<data> value_data_fut;
              };

              auto state = std::make_shared<state_>();
              state->key_data_fut = to_data2(key);
              state->value_data_fut = to_data2(value);

              auto result = state->key_data_fut
                  .then(launch::sync, [state](boost::future<data> f){
                    // store the ready key_data_fut for later
                    state->key_data_fut = std::move(f);
                    return std::move(state->value_data_fut);
                  })
                  .unwrap()
                  .then(launch::sync, [state, ttl, this](boost::future<data> f){
                    auto key_data = state->key_data_fut.get();
                    auto value_data = f.get();
                    return put_internal(key_data, value_data, ttl);
                  })
                  .unwrap()
                  .then(launch::sync, [this](boost::future<boost::optional<data>> f){
                    return to_object2<R>(f.get());
                  })
                  .unwrap();

              return result;
```